[RELAND] Close some sources of fake tensor leakage

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161589

Reland of https://github.com/pytorch/pytorch/pull/159923 

Couple of fixes:
1. When we run into an operation we didn't proxy, we end up emitting fake constants. We detect this and warn using the FQN of the lifted constant. We warn because some internal users complained it was regressing their exportability. 

2. Previous attribute mutation detection logic in non-strict didn't account for nested module structure. This fixes silent incorrectness issue of exporting esm and qwen in non-strict

3. We modify yolov3 to fix the previous silent incorrect behaviour 
4. We use strict export for levit_128 because it errors in non-strict due to more strict side effect checking 


When upgrading torchbench pin, opacus_cifar10 seems to not run on eager anymore. I verified this by pushing a temporary PR on master with new pin. So i added it to expect_fail list. 



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

Differential Revision: [D81326893](https://our.internmc.facebook.com/intern/diff/D81326893)

tugsbayasgalan committed:
[RELAND] Close some sources of fake tensor leakage

[ghstack-poisoned]

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -420,6 +420,28 @@ def forward(self, x):
+                return self.bar(x)
+
+        foo = Foo(Bar())
+        with self.assertRaisesRegex(ValueError, "self.bar._cache"):
```
@avikchaudhuri I am bit unsure what the correct behaviour here should be. Should we just warn to be consistent with other checks I have in this PR? If we decide to just warn whenever something sketchy happens, we could relax https://github.com/pytorch/pytorch/blob/8b78ba07b14d774f889f3b863a766e65d60cabff/torch/_functorch/_aot_autograd/frontend_utils.py#L280 these ones as well. Wdyt?

tugsbayasgalan replied:
It will just capture the tensor computations so bar._cache and operations on them won't show up. So we are kind of dropping side effects.

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -2027,6 +2035,43 @@ def _produce_guards_callback(gm):
+            if user.meta.get("stack_trace", None) is not None:
+                dependencies.append(user.meta["stack_trace"])
+        if len(dependencies) > 0:
+            warnings.warn(
```
You mentioned in the summary that some internal users complained, but isn't this a pretty bad situation to run into?

Can you think of a case where tracing was done with fake tensors that could potentially cause this situation to be hit?

tugsbayasgalan replied:
Yeah in their case, constants didn't show up in the graph so it was kinda okay. We still need to get repro from them and fix it properly but they were pretty persistent about it being not fatal error for them.

tugsbayasgalan replied:
ok after thinking about it bit more, i think this situation is almost always gonna be pretty bad. We should probably just error and disable the error on their side. I feel it would be good to let ppl complain to us when they hit this error, so that we actually find corner cases in our tracing system.

avikchaudhuri reviewed (commented):
some questions...

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1852,6 +1852,14 @@ def _find_node(gm: torch.fx.GraphModule, name: str) -> torch.fx.Node:
+    if len(splitted_names) < 1:
+        return True
+
+    return splitted_names[-1].startswith("lifted_tensor")
```
Yeah sure. lifted_tensor usually is reserved for cases when we see a random tensor without any proxy (aka we failed to capture an operation on it)

tugsbayasgalan committed:
Update on "[RELAND] Close some sources of fake tensor leakage"


Reland of https://github.com/pytorch/pytorch/pull/159923 

Couple of fixes:
1. When we run into an operation we didn't proxy, we end up emitting fake constants. We detect this and warn using the FQN of the lifted constant. We warn because some internal users complained it was regressing their exportability. 

2. Previous attribute mutation detection logic in non-strict didn't account for nested module structure. This fixes silent incorrectness issue of exporting esm and qwen in non-strict

3. We modify yolov3 to fix the previous silent incorrect behaviour 
4. We use strict export for levit_128 because it errors in non-strict due to more strict side effect checking 


When upgrading torchbench pin, opacus_cifar10 seems to not run on eager anymore. I verified this by pushing a temporary PR on master with new pin. So i added it to expect_fail list. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[RELAND] Close some sources of fake tensor leakage"


Reland of https://github.com/pytorch/pytorch/pull/159923 

Couple of fixes:
1. When we run into an operation we didn't proxy, we end up emitting fake constants. We detect this and warn using the FQN of the lifted constant. We warn because some internal users complained it was regressing their exportability. 

2. Previous attribute mutation detection logic in non-strict didn't account for nested module structure. This fixes silent incorrectness issue of exporting esm and qwen in non-strict

3. We modify yolov3 to fix the previous silent incorrect behaviour 
4. We use strict export for levit_128 because it errors in non-strict due to more strict side effect checking 


When upgrading torchbench pin, opacus_cifar10 seems to not run on eager anymore. I verified this by pushing a temporary PR on master with new pin. So i added it to expect_fail list. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela 

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D81133908).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D81133908).

atalman commented:
@pytorchmergebot revert -c nosignal -m "[GH job link](https://github.com/pytorch/pytorch/actions/runs/17305150611/job/49128381649) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/5790b009751e6ebba35d3e6d05e7c1b135553eee)"

atalman commented:
@tugsbayasgalan unlanding internal changes as well

huydhn commented:
@tugsbayasgalan I'm bumping TorchBench commit hash at the moment in https://github.com/pytorch/pytorch/pull/161461, which includes your change there.  Let me know if you want to coordinate to land this to avoid missing signals from inductor jobs

tugsbayasgalan committed:
Update on "[RELAND] Close some sources of fake tensor leakage"


Reland of https://github.com/pytorch/pytorch/pull/159923 

Couple of fixes:
1. When we run into an operation we didn't proxy, we end up emitting fake constants. We detect this and warn using the FQN of the lifted constant. We warn because some internal users complained it was regressing their exportability. 

2. Previous attribute mutation detection logic in non-strict didn't account for nested module structure. This fixes silent incorrectness issue of exporting esm and qwen in non-strict

3. We modify yolov3 to fix the previous silent incorrect behaviour 
4. We use strict export for levit_128 because it errors in non-strict due to more strict side effect checking 


When upgrading torchbench pin, opacus_cifar10 seems to not run on eager anymore. I verified this by pushing a temporary PR on master with new pin. So i added it to expect_fail list. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D81133908](https://our.internmc.facebook.com/intern/diff/D81133908)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D81326893).

tugsbayasgalan committed:
Update on "[RELAND] Close some sources of fake tensor leakage"


Reland of https://github.com/pytorch/pytorch/pull/159923 

Couple of fixes:
1. When we run into an operation we didn't proxy, we end up emitting fake constants. We detect this and warn using the FQN of the lifted constant. We warn because some internal users complained it was regressing their exportability. 

2. Previous attribute mutation detection logic in non-strict didn't account for nested module structure. This fixes silent incorrectness issue of exporting esm and qwen in non-strict

3. We modify yolov3 to fix the previous silent incorrect behaviour 
4. We use strict export for levit_128 because it errors in non-strict due to more strict side effect checking 


When upgrading torchbench pin, opacus_cifar10 seems to not run on eager anymore. I verified this by pushing a temporary PR on master with new pin. So i added it to expect_fail list. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D81326893](https://our.internmc.facebook.com/intern/diff/D81326893)

[ghstack-poisoned]
